src/library_filters.c

```diff
@@ -15,6 +15,8 @@ Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 #include "library_filters.h"
 #include "libc_wrappers.h"
 
+#include "gotcha/gotcha.h"
+
 static const char* filter;
```

Comment:
I think that filter should be thread local with an __thread declaration to provide safety in multithreaded programs.

Reply:
No, the filters are only used under the gotcha_wrap call (which is not thread safe, and makes no promises to be).  They're not used during the actual call, when we would need thread safety.

Reply:
Does it make sense to allow each thread to perform different filtering? 

I feel the current interface assumes that library filter is only set before spawning threads and is applied to all threads.

Reply:
Hmm. But exposing gotcha_filter_libraries_by_name as an external function, breaks Matt's invariant of "called under only gotcha_wrap".

Reply:
Sorry, I should be more clear.

There's an "init" time where you setup filters, do configuration, and call gotcha_wrap().  That entire "init" period should be run with a single thread, and gotcha makes no attempts at thread safety during those calls.  If you have multiple threads that might trigger a gotcha "init", then put mutexes external to gotcha around that.

Then there's the "call" time, when the app is executing and bouncing to the wrappers.  Anything there should absolutely be thread safe.  But these filters aren't accessed at that time.

Reply:
Any library might want to employ gotcha. libhpcrun.so is just a library. So when someone dlopens libhpcrun.so, our init constructors can call gotcha APIs. 

I hate to be the bearer of bad news, but any thread can dlopen a library. To avoid conflicts when multiple concurrently-loaded libraries try to use gotcha, it seems like gotcha should supply a lock that a client library could acquire before invoking gotcha operations and release after they are done. That would be a minimal requirement to enable a client of gotcha to work in a fashion that can be guaranteed to be thread safe. It would also require little of the gotcha implementation other than the requirement that the only function it is safe to call without holding the gotcha lock is gotcha_lock_aquire(). All other functions may only be called if the gotcha lock is held. You don't have to put in code to enforce that. A stern warning in the documentation should be enough.

Having a lock separate from gotcha provides no safety from gotcha usage by init constructors from concurrently loaded libraries.

Reply:
That's a concern, but not something we should address in this PR.